src/main/scala/linting/rule/RenameDesiredNames.scala

```diff
@@ -0,0 +1,202 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.linting
+package rule
+
+import firrtl._
+import firrtl.annotations.{CircuitTarget, HasSerializationHints, IsModule, SingleTargetAnnotation, Target}
+import firrtl.ir._
+import firrtl.options.Dependency
+import firrtl.transforms.DedupModules
+
+import scala.collection.mutable
+
+/** A helper to rename modules in a [[Circuit]]
+  */
+object RenameModules {
+  def onStmt(moduleNameMap: Map[String, String])(stmt: Statement): Statement = stmt match {
+    case inst: WDefInstance if moduleNameMap.contains(inst.module) => inst.copy(module = moduleNameMap(inst.module))
+    case inst: DefInstance if moduleNameMap.contains(inst.module) => inst.copy(module = moduleNameMap(inst.module))
+    case other => other.mapStmt(onStmt(moduleNameMap))
+  }
+
+  /** Renames the modules in a circuit given a mapping of old names to new names
+    *
+    * @param nameMappings mapping of old to new names
+    * @param circuit the circuit to rename
+    */
+  def apply(nameMappings: Map[String, String], circuit: Circuit): Circuit = {
+    val modules = circuit.modules.map {
+      case mod: Module => mod.mapStmt(onStmt(nameMappings)).mapString(m => nameMappings.getOrElse(m, m))
+      case ext: ExtModule => ext
+    }
+    val main = nameMappings.getOrElse(circuit.main, circuit.main)
+    circuit.copy(main = main, modules = modules)
+  }
+}
+
+/** A strategy for generating new names for modules
+  */
+sealed trait NamingStrategy {
+
+  /** Generates a new stable module name based on the desired name and the module IR node
+    *
+    * called by [[RenameDesiredNames]] to rename modules
+    *
+    * @param desiredName the requested name for the module; generated name should contain this name
+    * @param module the module targeted by the desiredName
+    */
+  def getName(desiredName: String)(module: Module): String
+}
+
+/** A naming strategy that always returns the desired name
```

Comment:
This isn't technically true. It relies on the linting being run now to 'always' return the desired name. I think this should be explicit that it either exactly renames or triggers a linting error (if linting is run).